File: fe/src/api/processing.ts
import { apiFetch } from './client'

export interface ProcessingItem {
  index_id: number
  batch_id: number
  batch_index_id: number
  text_from: string
  text_to: string
  line_id_from: string // JSON array string e.g. "[1,2]"
  line_id_to: string // JSON array string e.g. "[3,4]"
  processing_from_id: number
  processing_to_id: number
}

export interface ProcessingResponse {
  items: ProcessingItem[]
  meta: {
    page: number
    total_pages: number
  }
  proxy_from_dict: Record<string, unknown>
  proxy_to_dict: Record<string, unknown>
}

export interface MetaResponse {
  total_items: number
  lang_from: string
  lang_to: string
}

export interface IndexEntry {
  id: number
  index_id: number
  text_type: string
  text: string
}

export interface IndexResponse {
  index: IndexEntry[]
}

export interface EditParams {
  index_id: number
  text: string
  text_type: string
  operation: string
  batch_id: number
  batch_index_id: number
  target?: string
  candidate_line_id?: number
  candidate_text?: string
  line_id_from?: number
  line_id_to?: number
}

export interface SplitSentenceParams {
  direction: string
  line_id: number
  part1: string
  part2: string
}

export interface CandidateLine {
  id: number
  index_id: number
  text: string
}

export interface CandidatesResponse {
  items: CandidateLine[]
}

export function getProcessingPage(guid: string, count: number, page: number) {
  return apiFetch<ProcessingResponse>(
    `/api/aligner/processing/${guid}/page?count=${count}&page=${page}`,
  )
}

export function getProcessingMeta(guid: string) {
  return apiFetch<MetaResponse>(`/api/aligner/processing/${guid}/meta`)
}

export function getDocIndex(guid: string) {
  return apiFetch<IndexResponse>(`/api/aligner/processing/${guid}/index`)
}

export function editProcessing(guid: string, data: EditParams) {
  return apiFetch<void>(`/api/aligner/processing/${guid}/edit`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function splitSentence(guid: string, data: SplitSentenceParams) {
  return apiFetch<void>(`/api/aligner/processing/${guid}/split`, {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function getCandidates(
  guid: string,
  textType: string,
  indexId: number,
  countBefore: number,
  countAfter: number,
  shift: number,
) {
  const params = new URLSearchParams({
    text_type: textType,
    index_id: String(indexId),
    count_before: String(countBefore),
    count_after: String(countAfter),
    shift: String(shift),
  })
  return apiFetch<CandidatesResponse>(
    `/api/aligner/processing/${guid}/candidates?${params}`,
  )
}

export function switchExcluded(guid: string, lineId: number, textType: string) {
  const params = new URLSearchParams({
    line_id: String(lineId),
    text_type: textType,
  })
  return apiFetch<void>(`/api/aligner/processing/${guid}/exclude?${params}`, {
    method: 'POST',
  })
}

File: be/app/routers/processing.py
"""Processing router - aligned pairs, editing, splitting, candidates"""

import logging

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.user import User
from app.schemas.alignment import EditRequest, SplitRequest
from app.services import alignment_service, editor_service

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/aligner/processing", tags=["processing"])


@router.get("/{guid}/page")
def get_processing_page(
    guid: str,
    count: int = Query(50),
    page: int = Query(1),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    return editor_service.get_processing_page(user.id, alignment, count, page)


@router.post("/{guid}/page/by-ids")
def get_processing_by_ids(
    guid: str,
    index_ids: list[int],
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    return editor_service.get_processing_by_ids(user.id, alignment, index_ids)


@router.get("/{guid}/meta")
def get_processing_meta(
    guid: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    return {"meta": editor_service.get_processing_meta(user.id, alignment)}


@router.get("/{guid}/index")
def get_doc_index(
    guid: str,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    return {"items": editor_service.get_doc_index(user.id, alignment)}


@router.post("/{guid}/edit")
def edit_processing(
    guid: str,
    data: EditRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")
    editor_service.edit_doc(user.id, alignment, data)
    return {"status": "ok"}


@router.post("/{guid}/split")
def split_sentence(
    guid: str,
    data: SplitRequest,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    if data.direction not in ("from", "to"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid direction"
        )

    editor_service.split_sentence(user.id, alignment, data)
    return {"status": "ok"}


@router.get("/{guid}/candidates")
def get_candidates(
    guid: str,
    text_type: str = Query("to"),
    index_id: int = Query(...),
    count_before: int = Query(10),
    count_after: int = Query(10),
    shift: int = Query(0),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    if text_type not in ("from", "to"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid text_type"
        )

    candidates = editor_service.get_candidates(
        user.id, alignment, text_type, index_id, count_before, count_after, shift
    )
    return {"items": candidates}


@router.post("/{guid}/exclude")
def switch_excluded(
    guid: str,
    line_id: int = Query(...),
    text_type: str = Query("from"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    editor_service.switch_excluded(user.id, alignment, line_id, text_type)
    return {"status": "ok"}


@router.get("/{guid}/splitted/{direction}")
def get_splitted_by_ids(
    guid: str,
    direction: str,
    ids: str = Query("[]"),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    if direction not in ("from", "to"):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid direction"
        )

    import json
    try:
        id_list = json.loads(ids)
    except Exception:
        id_list = []

    return {"items": editor_service.get_splitted_by_ids(user.id, alignment, direction, id_list)}


@router.get("/{guid}/find/{lang}/{line_id}")
def get_line_position(
    guid: str,
    lang: str,
    line_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    alignment = alignment_service.get_alignment(db, user.id, guid)
    if not alignment:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Alignment not found")

    pos = editor_service.get_line_position(user.id, alignment, lang, line_id)
    return {"pos": pos}
